File: high_perf_python/wordcount.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s filename\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    int ch;
    int word_count = 0;
    int in_word = 0;

    while ((ch = fgetc(file)) != EOF) {
        if (isspace(ch)) {
            if (in_word) {
                in_word = 0;
                word_count++;
            }
        } else {
            in_word = 1;
        }
    }

    // If the last character was part of a word, we need to count that word.
    if (in_word) {
        word_count++;
    }

    fclose(file);
    printf("Number of words: %d\n", word_count);

    return 0;
}

File: high_perf_python/wordcount.py
import sys
from io import TextIOWrapper

def count_words(filename):
    try:
        word_count = 0
        in_word = False

        with open(filename, 'r', encoding='utf-8') as file:
            for line in file:
                for ch in line:
                    if isspace(ch):
                        if in_word:
                            in_word = False
                            word_count += 1
                    else:
                        in_word = True

                # Handle the case where the line ends with a word
                if in_word:
                    word_count += 1
                    in_word = False

        return word_count
    except FileNotFoundError:
        print(f"Error: The file {filename} was not found.")
        return None
    except IOError:
        print(f"Error: Could not read the file {filename}.")
        return None

def isspace(ch):
    return ch in ' \t\n\r'

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python wordcount.py filename")
        sys.exit(1)

    filename = sys.argv[1]
    word_count = count_words(filename)

    if word_count is not None:
        print(f"Number of words: {word_count}")